### ISSUE TITLE
[Feature] Possibility to load and execute hl bytecode in runtime

### ISSUE THREAD

**Issue by Grabli66** [HaxeFoundation/hashlink#179]

**Comment by ncannasse** [comment]
This is a bit tricky, as ATM we only have one single "module" that can be loaded. However this should only affect very specific parts of the runtime, mostly requiring to change `module.c` to keep and search into a list of modules, as well as changes in `debugger.c` to support multiple modules debugging.

Exposing then a load_module function would be doing pretty much what `main.c` is doing already (reading the file, loading the bytecode, jitting it and calling the main entry point).

The main question is about what kind of interoperability/communications we want between the two otherwise completely isolated modules.

**Comment by Grabli66** [comment]
I see it like a possibility to load and use types from other module, to create plugins, extensions. Like this(pseudo haxe code):

import hl.ModuleLoader;

interface IPlugin {
  function myMethod()  : Int;
}

var module = ModuleLoader.loadModule("extension.hl");
var myPlugin = cast(new module.Plugin(), IPlugin);
trace(myPlugin.myMethod());

But if modules are isolated, than i think there will be a problems with shared types between that modules, cause it will be different types in each module. But i am not sure :)

**Comment by ncannasse** [comment]
Yes exactly, in that case both plugin and core module would have their own IPlugin type, one not matching the other. Thinking about it, all required would be to remap all loaded module types to loader types in case they match exactly (same name, same fields, etc.). Definitely feasible but requires a bit of extra work.

**Comment by gdeverlant** [comment]
How about thinking of something that was working really well inside of Adobe Flash/AIR with the Loader class. Doing something like this with ApplicationDomain and dynamic resources loading. We can inspire ourselves with what worked really well.

**Comment by ncannasse** [comment]
Some progress has been made in recent changes, as we now allow several JIT modules to be loaded.

**Comment by sonygod** [comment]
any progress? Are there any example to show this feature? @ncannasse

**Comment by gdeverlant** [comment]
I will like to port my OSGi runtime framework from Actionscript to Haxe but for that I need ApplicationDomain and Loader. The ability to Loader.load() and Loader.unload() will allow me to bring hot code swapping.

Right now it works really well in Actionscript 3.

**Comment by dmitryhryppa** [comment]
Any progress on this?
Would be really nice to provide a plugins-like system. It may be very useful for modding games made with Heaps :)

**Comment by hakanrw** [comment]
i also believe this would be useful. it would be great if hashlink runtime could load more than one hl file. similar to how JVM/CLR can load all files in the classpath.

unrelated to hashlink, but an additional flag in the haxe compiler named `--hl-lib` similar to `--java-lib` and `--net-lib` would make sense in this scenario. this way, haxe libraries can be distributed as hashlink files, to be used during compilation.

**Comment by posxposy** [comment]
Look at thaaaat 👀
https://github.com/HaxeFoundation/haxe/commit/e4b21dbfd82b6b7bea48ec1c729b6db5a67e4f0e

**Comment by ncannasse** [comment]
Hashlink JIT now supports runtime code loading using hl.Api.loadPlugin("file.hl"). Requires GIT Haxe and HL atm.

You can share classes between the host and the plugin by adding @:virtual to the class declaration then --macro exclude("ClassName") when compiling the plugin which will load it dynamically from the host.

**Comment by nathanvoglsam** [comment]
Just curious, how does this interact with HL/C?

Would I be able to, say, pre-compile a core engine API module with HL/C and then dynamically load .hl modules that import the core API using the HL/C code?

Super excited to see this going in!

**Comment by ncannasse** [comment]
This is not supported by HL/C atm.
